BrowserKit/Sources/TabDataStore/TabDataStore.swift

```diff
@@ -11,11 +11,53 @@ protocol TabDataStore {
 }
 
 actor DefaultTabDataStore: TabDataStore {
+    private var lastSaveTime: Date = .distantPast
+    private let minimumSaveInterval: TimeInterval = 5
+
+    private let filePath: URL = {
+        let fileManager = FileManager.default
+        let documentsDirectory = try? fileManager.url(for: .documentDirectory,
+                                                      in: .userDomainMask,
+                                                      appropriateFor: nil,
+                                                      create: false)
+        return documentsDirectory?.appendingPathComponent("WindowData.json") ?? URL(fileURLWithPath: "")
+    }()
+
     func fetchTabData() async -> WindowData {
-        return WindowData(id: UUID(), isPrimary: true, activeTabId: UUID(), tabData: [])
+        do {
+            let data = try Data(contentsOf: filePath)
+            let windowData = try JSONDecoder().decode(WindowData.self, from: data)
+            return windowData
+        } catch {
+            return WindowData(id: UUID(), isPrimary: true, activeTabId: UUID(), tabData: [])
+        }
     }
 
-    func saveTabData(window: WindowData) async {}
+    func saveTabData(window: WindowData) async {
+        let now = Date()
+        guard now.timeIntervalSince(lastSaveTime) >= minimumSaveInterval else { return }
+        lastSaveTime = now
 
-    func clearAllTabData() async {}
+        do {
+            let data = try JSONEncoder().encode(window)
+            let url = try FileManager.default.url(for: .documentDirectory,
+                                                  in: .userDomainMask,
+                                                  appropriateFor: nil,
+                                                  create: false)
+                .appendingPathComponent("WindowData.json")
+            try data.write(to: url)
+        } catch {
+            print("Error saving tab data: \(error)")
```

Comment:
We don't allow prints to be merged to main. You can instead inject the logger and log an error.